beebop/assignClusters.py

```diff
@@ -27,93 +33,421 @@ def hex_to_decimal(sketches_dict) -> None:
+        to external cluster labels]
+    :param previous_query_clustering: [path to previous
+        query clustering file]
+    """
+    not_found_prev_querying = config.fs.external_previous_query_clustering_tmp(
+        config.p_hash
+    )
+    external_clusters_not_found, _ = get_external_clusters_from_file(
+        not_found_prev_querying,
+        not_found_query_names,
+        config.external_clusters_prefix,
+    )
+    update_external_clusters_csv(
+        previous_query_clustering,
+        not_found_query_names,
+        external_clusters_not_found,
+    )
+    external_clusters.update(external_clusters_not_found)
+
+
+def merge_partial_query_graphs(p_hash: str, fs: PoppunkFileStore) -> None:
+    """
+    [Merge the partial query graphs from the full assign job with the
+    partial query graphs from the initial assign job.]
+
+    :param p_hash: [project hash]
+    :param fs: [PoppunkFileStore with paths to input files
+    """
+    tmp_assign_subset_file = fs.partial_query_graph_tmp(p_hash)
+    main_subset_file = fs.partial_query_graph(p_hash)
+    with open(tmp_assign_subset_file, "r") as f:
+        failed_lines = set(f.read().splitlines())
+    with open(main_subset_file, "r") as f:
+        main_lines = set(f.read().splitlines())
+
+    combined_lines = list(main_lines.union(failed_lines))
+    with open(main_subset_file, "w") as f:
+        f.write("\n".join(combined_lines))
+
+
+def copy_include_files(output_full_tmp: str, outdir: str) -> None:
+    """
+    [Copy include files from the full assign output directory
+        to the output directory
+    where all files from the PopPUNK assign job are stored.]
+
+    :param output_full_tmp: [path to full assign output directory]
+    :param outdir: [path to output directory]
+    """
+    include_files = [
+        f for f in os.listdir(output_full_tmp) if f.startswith("include")
+    ]
+    for include_file in include_files:
+        os.rename(
+            f"{output_full_tmp}/{include_file}", f"{outdir}/{include_file}"
         )
-    else:
-        for i, (name, cluster) in enumerate(
-            zip(queries_names, queries_clusters)
-        ):
-            result[i] = {
-                "hash": name,
-                "cluster": cluster
-            }
-
-    # save result to retrieve when reloading project results - this
-    # overwrites the initial output file written before the assign
-    # job ran
-    with open(fs.output_cluster(p_hash), 'wb') as f:
-        pickle.dump(result, f)
 
+
+def filter_queries(
+    queries_names: list[str],
+    queries_clusters: list[str],
+    not_found: list[str],
+    config: ClusteringConfig,
+) -> tuple[list[str], list[str]]:
+    """
+    [Filter out queries that were not found in the
+        initial external clusters file.
+    This function filters out the queries that were not found
+        in the initial external clusters file,
+    deletes include files for these queries,
+        and returns the filtered queries.]
```

Comment:
it feels like a little bit of an odd hybrid for this method to do filtering on lists for use by the caller, and also to delete some files. Maybe split these parts up?

Reply:
agreed